packages/xo-server/src/xo-mixins/metadata-backups.js

```diff
@@ -0,0 +1,247 @@
+    deleteSchedule: (id: string) => Promise<void>,
+    getXapi: (id: string) => Xapi,
+    getJob: (
+      id: string,
+      ?METADATA_BACKUP_JOB_TYPE
+    ) => Promise<MetadataBackupJob>,
+    updateJob: (
+      $Shape<MetadataBackupJob>,
+      ?boolean
+    ) => Promise<MetadataBackupJob>,
+    removeJob: (id: string) => Promise<void>,
+  }
+
+  constructor(app: any) {
+    this._app = app
+    app.on('start', () => {
+      app.registerJobExecutor(
+        METADATA_BACKUP_JOB_TYPE,
+        this._executor.bind(this)
+      )
+    })
+  }
+
+  @deferrable
+  async _executor($defer, { cancelToken, job: job_, schedule }): Executor {
+    if (schedule === undefined) {
+      throw new Error('backup job cannot run without a schedule')
+    }
+
+    const job: MetadataBackupJob = (job_: any)
+    const remoteIds = unboxIdsFromPattern(job.remotes)
+    if (remoteIds.length === 0) {
+      throw new Error('metadata backup job cannot run without remotes')
+    }
+
+    const poolIds = unboxIdsFromPattern(job.pools)
+    const isEmptyPools = poolIds.length === 0
+    if (!job.xoMetadata && isEmptyPools) {
+      throw new Error('no metadata mode found')
+    }
+
+    const app = this._app
+    const { retentionXoMetadata, retentionPoolMetadata } =
+      job?.settings[schedule.id] || {}
+
+    const scheduleDir = `${METADATA_BACKUP_DIR}/${schedule.id}`
+    const timestamp = safeDateFormat(Date.now())
+    const files = []
+    if (job.xoMetadata && retentionXoMetadata > 0) {
+      const dir = `${scheduleDir}/xo`
+      const fileName = `${dir}/${timestamp}.json`
+      const metaDataFileName = `${dir}/${timestamp}.metadata.json`
+
+      const data = await app.exportConfig()
+
+      files.push({
+        executeBackup: async handler => {
+          await Promise.all([
+            handler.outputFile(fileName, JSON.stringify(data, null, 2)),
+            handler.outputFile(
+              metaDataFileName,
+              JSON.stringify(
+                {
+                  jobId: job.id,
+                  scheduleId: schedule.id,
+                  timestamp,
+                },
+                null,
+                2
+              )
+            ),
+          ])
+          $defer.onFailure(async () => {
+            await Promise.all([
+              handler.unlink(fileName),
+              handler.unlink(metaDataFileName),
+            ])
+          })
+        },
+        dir,
+        retention: retentionXoMetadata - 1,
+      })
+    }
+    if (!isEmptyPools && retentionPoolMetadata > 0) {
+      const retention = retentionPoolMetadata - 1
+      files.push(
+        ...(await Promise.all(
+          poolIds.map(async id => {
+            const dir = `${scheduleDir}/${id}`
+            const fileName = `${dir}/${timestamp}.xml`
+            const metaDataFileName = `${dir}/${timestamp}.metadata.json`
+
+            const stream = await app.getXapi(id).getPoolMetadata(cancelToken)
```

Comment:
It would be nice to export the metadata only once and then split the stream, but it's hard to get right.

I think you should add a `TODO` comment above for now.